File: frontend/src/api/externalSystem.ts
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import apiClient from './client'

/** sso_link = 同 SSO 应用（与 KkOps 共用同一 IdP，打开即已登录）；jwt_token = Token 跳转（附带身份与权限） */
export type LaunchType = 'sso_link' | 'jwt_token'

export interface ExternalSystem {
  id: number
  name: string
  description?: string
  launch_type: LaunchType
  base_url: string
  login_path?: string
  role_mapping?: string
  icon?: string
  order_index: number
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface CreateExternalSystemRequest {
  name: string
  description?: string
  launch_type?: LaunchType
  base_url: string
  login_path?: string
  secret?: string
  role_mapping?: string
  icon?: string
  order_index?: number
  enabled?: boolean
}

export interface UpdateExternalSystemRequest {
  name?: string
  description?: string
  launch_type?: LaunchType
  base_url?: string
  login_path?: string
  secret?: string
  role_mapping?: string
  icon?: string
  order_index?: number
  enabled?: boolean
}

export interface LaunchResponse {
  redirect_url: string
}

export const externalSystemApi = {
  list: (enabledOnly = true) =>
    apiClient.get<{ data: ExternalSystem[] }>(
      `/external-systems${enabledOnly ? '' : '?enabled=false'}`
    ),
  get: (id: number) =>
    apiClient.get<{ data: ExternalSystem }>(`/external-systems/${id}`),
  create: (data: CreateExternalSystemRequest) =>
    apiClient.post<{ data: ExternalSystem }>('/external-systems', data),
  update: (id: number, data: UpdateExternalSystemRequest) =>
    apiClient.put<{ data: ExternalSystem }>(`/external-systems/${id}`, data),
  delete: (id: number) =>
    apiClient.delete<unknown>(`/external-systems/${id}`),
  launch: (id: number) =>
    apiClient.post<LaunchResponse>(`/external-systems/${id}/launch`),
}

File: backend/internal/model/external_system.go
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package model

import (
	"time"

	"gorm.io/gorm"
)

// LaunchType: sso_link = same SSO (e.g. Keycloak), just open URL; jwt_token = KkOps issues JWT with user/perms
const (
	LaunchTypeSSOLink  = "sso_link"  // 同 SSO 应用：与 KkOps 共用同一 IdP，打开即已登录
	LaunchTypeJWTToken = "jwt_token" // Token 跳转：目标系统接收 KkOps 签发的 JWT
)

// ExternalSystem represents an external ops system for SSO portal (open after one login)
type ExternalSystem struct {
	ID          uint   `gorm:"primaryKey" json:"id"`
	Name        string `gorm:"not null;size:100" json:"name"`
	Description string `gorm:"type:text" json:"description"`
	// LaunchType: sso_link = open base_url only (same IdP); jwt_token = build JWT and redirect
	LaunchType string `gorm:"size:32;default:jwt_token" json:"launch_type"`
	// For both: base_url (sso_link = full URL to open; jwt_token = scheme+host)
	BaseURL string `gorm:"not null;size:512" json:"base_url"`
	// For jwt_token only: path that accepts token, e.g. /api/sso/consume
	LoginPath   string         `gorm:"size:255" json:"login_path"`
	Secret      string         `gorm:"size:255" json:"-"` // for jwt_token: shared secret
	RoleMapping string         `gorm:"type:text" json:"role_mapping"`
	Icon        string         `gorm:"size:64" json:"icon"`
	OrderIndex  int            `gorm:"default:0" json:"order_index"`
	Enabled     bool           `gorm:"default:true" json:"enabled"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName specifies table name
func (ExternalSystem) TableName() string {
	return "external_systems"
}
